sdk/log/batch.go

```diff
@@ -76,6 +78,88 @@ func (b *BatchingProcessor) ForceFlush(ctx context.Context) error {
 	return nil
 }
 
+// queue holds a queue of logging records.
+//
+// When the queue becomes full, the oldest records in the queue are
+// overwritten.
+type queue struct {
+	sync.Mutex
+
+	cap, len    int
+	read, write *ring.Ring
+}
+
+func newQueue(size int) *queue {
+	r := ring.New(size)
+	return &queue{
+		cap:   size,
+		read:  r,
+		write: r,
+	}
+}
+
+// Enqueue adds r to the queue. The queue size, including the addition of r, is
+// returned.
+//
+// If enqueueing r will exceed the capacity of q, the oldest Record held in q
+// will be dropped and r retained.
+func (q *queue) Enqueue(r Record) int {
+	q.Lock()
+	defer q.Unlock()
+
+	q.write.Value = r
+	q.write = q.write.Next()
+
+	q.len++
+	if q.len > q.cap {
+		// Overflow. Advance read to be the new "oldest".
+		q.len = q.cap
+		q.read = q.read.Next()
+	}
+	return q.len
+}
+
+// TryDequeue attempts to dequeue up to len(buf) Records. The available Records
+// will be assigned into buf and passed to write. If write fails, returning
+// false, the Records will not be removed from the queue. If write succeeds,
+// returning true, the dequeued Records are removed from the queue. The number
+// of Records remaining in the queue are returned.
+func (q *queue) TryDequeue(buf []Record, write func([]Record) bool) int {
+	q.Lock()
+	defer q.Unlock()
+
+	origRead := q.read
+
+	n := min(len(buf), q.len)
+	for i := 0; i < n; i++ {
+		buf[i] = q.read.Value.(Record)
+		q.read = q.read.Next()
+	}
+
+	if write(buf[:n]) {
```

Comment:
An academical type of comment...

Calling external code (`write`) when a lock is acquired can lead to a deadlock. I was thinking that maybe we could pass the `write` handler to `newQueue` to reduce the possibility of having a deadlock?

Feel free to resolve if you think that discussing is a waste of time.

Reply:
I don't follow.

The `write` function sends to a chan if there is space. There is no locking there.

How would passing the `bufferExporter` to `queue` change that? Wouldn't it just change the call signature and couple the two types? You would still need to hold the lock given you don't know yet if you can release the data or not yet.

Reply:
> How would passing the `bufferExporter` to queue change that? 

It would reduce the possibility that someone e.g. calls `q.Flush` in `write`. As mentioned, it is almost an academical type of comment. Reference: https://github.com/isocpp/CppCoreGuidelines/issues/549

Reply:
Gotcha. I think, like you mention, this is all unexported and we can keep it the way it is.

I'll add a comment to help OTel devs understand that `write` should not call the `queue`.